File: backend/tiramisu-lite/Model/PlanerItem.cs
namespace tiramisu_lite.Model;

public class PlanerItem
{
    public Guid Id { get; set; }
    public Guid PlanerId { get; set; }
    public Planer Planer { get; init; } = null!;
    public string Title { get; set; }
    public DateTime EatTime { get; set; }
    public bool Notify { get; set; }

    public required IEnumerable<Meal> Meals
    {
        get => this.meals;
        init => this.meals = value.ToList();
    }

    public decimal KcalSummary => this.meals.Sum(s => s.Kcal);
    private List<Meal> meals = [];

    public PlanerItem(Guid id, Guid planerId, string title, DateTime eatTime, bool notify = false)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(title);
        this.Id = id;
        this.PlanerId = planerId;
        this.Title = title;
        this.EatTime = eatTime;
        this.Notify = notify;
    }


    public void UpdateTitle(string title)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(title, nameof(title));
        this.Title = title;
    }

    public void UpdateEatTime(DateTime eatTime)
    {
        this.EatTime = eatTime;
    }

    public void UpdateNotify(bool notify)
    {
        this.Notify = notify;
    }
}

File: mobile/lib/model/planer_item.dart
import 'package:mobile/model/meal.dart';

class PlanerItem {
  final String id;
  final String title;
  final DateTime eatTime;
  final List<Meal> meals;
  final double kcalSummary;
  final bool notify;

  const PlanerItem({
    required this.id,
    required this.title,
    required this.eatTime,
    required this.meals,
    required this.kcalSummary,
    required this.notify,
  });

  factory PlanerItem.fromJson(Map<String, dynamic> json) {
    return PlanerItem(
      id: json['id'],
      title: json['title'],
      eatTime: DateTime.parse(json['eatTime']),
      meals:
          (json['meals'] as List).map((meal) => Meal.fromJson(meal)).toList(),
      kcalSummary: (json['kcalSummary']).toDouble(),
      notify: json['notify'],
    );
  }

  PlanerItem copyWith({
    String? id,
    String? title,
    DateTime? eatTime,
    List<Meal>? meals,
    double? kcalSummary,
    bool? notify,
  }) {
    return PlanerItem(
      id: id ?? this.id,
      title: title ?? this.title,
      eatTime: eatTime ?? this.eatTime,
      meals: meals ?? this.meals,
      kcalSummary: kcalSummary ?? this.kcalSummary,
      notify: notify ?? this.notify,
    );
  }
}
